libkbfs/delayed_cancellation_test.go

```diff
@@ -0,0 +1,135 @@
+		return context.WithValue(ctx, "test", "O_O")
+	})
+	ctx, cancel := context.WithCancel(ctx)
+
+	ctx, err := NewContextWithReplayFrom(ctx)
+	if err != nil {
+		t.Fatalf("calling NewContextWithReplayFrom error: %s", err)
+	}
+
+	// Test if replay was run properly
+	if ctx.Value("test") != "O_O" {
+		t.Fatalf("NewContextWithReplayFrom did not replay attached replayFunc")
+	}
+
+	// Test if cancellation is disabled
+	cancel()
+	select {
+	case <-ctx.Done():
+		t.Fatalf("NewContextWithReplayFrom did not disconnect the cancel function")
+	default:
+	}
+
+	// make sure the new ctx is also replayable
+	ctx, err = NewContextWithReplayFrom(ctx)
+	if err != nil {
+		t.Fatalf("calling NewContextWithReplayFrom error: %s", err)
+	}
+	if ctx.Value("test") != "O_O" {
+		t.Fatalf("NewContextWithReplayFrom did not replay attached replayFunc")
+	}
+}
+
+func makeContextWithDelayedCancellation(t *testing.T) (
+	ctx context.Context, originalCancel context.CancelFunc) {
+	ctx = context.Background()
+	ctx = NewContextReplayable(ctx, func(ctx context.Context) context.Context {
+		return context.WithValue(ctx, "test", "O_O")
+	})
+	ctx, cancel := context.WithCancel(ctx)
+
+	ctx, err := NewContextWithCancellationDelayer(ctx)
+	if err != nil {
+		t.Fatalf("calling NewContextWithCancellationDelayer error: %s", err)
+	}
+
+	// Test NewContextWithCancellationDelayer does replay properly
+	if ctx.Value("test") != "O_O" {
+		t.Fatalf(
+			"NewContextWithCancellationDelayer did not replay attached replayFunc")
+	}
+
+	return ctx, cancel
+}
+
+func TestDelayedCancellationCancelWhileNotEnabled(t *testing.T) {
+	t.Parallel()
+
+	ctx, cancel := makeContextWithDelayedCancellation(t)
+
+	cancel()
+
+	// wait until the go routine spawned in NewContextWithCancellationDelayer
+	// actually cancels
+	time.Sleep(10 * time.Millisecond)
+	select {
+	case <-ctx.Done():
+	default:
+		t.Fatalf("Cancellation did not happen even though " +
+			"EnableDelayedCancellationWithGracePeriod has not been called yet")
+	}
+}
+
+func TestDelayedCancellationCleanupWhileNotEnabled(t *testing.T) {
+	t.Parallel()
+
+	ctx, _ := makeContextWithDelayedCancellation(t)
+
+	if err := CleanupCancellationDelayer(ctx); err != nil {
+		t.Fatalf("calling CleanupCancellationDelayer error: %s", err)
+	}
+
+	// wait until the go routine spawned in NewContextWithCancellationDelayer
+	// actually cancels
+	time.Sleep(10 * time.Millisecond)
+	select {
+	case <-ctx.Done():
+	default:
+		t.Fatalf("Cancellation did not happen even though " +
+			"EnableDelayedCancellationWithGracePeriod has not been called yet")
+	}
+}
+
+func TestDelayedCancellationEnabled(t *testing.T) {
+	t.Parallel()
+
+	ctx, cancel := makeContextWithDelayedCancellation(t)
+	EnableDelayedCancellationWithGracePeriod(ctx, 15*time.Millisecond)
+
+	cancel()
+
+	// wait until any (presumably) wrong cancellation in the go routine spawned
+	// in NewContextWithCancellationDelayer actually cancels
+	time.Sleep(time.Millisecond * 10)
```

Comment:
This one doesn't really matter since `<-ctx.Done()` will block for 15 seconds anyway, but you can probably get rid of it too if you want, since it seems just as good to check `ctx.Done` immediately after the cancel to ensure that the context hasn't cancelled yet.